irksome/base_time_stepper.py

```diff
@@ -0,0 +1,155 @@
+import numpy
+
+from abc import abstractmethod
+from firedrake import Function, NonlinearVariationalProblem, NonlinearVariationalSolver
+from firedrake.dmhooks import pop_parent, push_parent
+from firedrake.petsc import PETSc
+from .tools import AI, get_stage_space, getNullspace
+from pyop2.types import MixedDat
+
+
+class BaseTimeStepper:
+    def __init__(self, F, t, dt, u0,
+                 bcs=None, appctx=None, nullspace=None):
+        self.F = F
+        self.t = t
+        self.dt = dt
+        self.u0 = u0
+        if bcs is None:
+            bcs = ()
+        self.orig_bcs = bcs
+        self.nullspace = nullspace
+        self.V = u0.function_space()
+
+        appctx_base = {
+            "F": F,
+            "t": t,
+            "dt": dt,
+            "u0": u0,
+            "bcs": bcs,
+            "nullspace": nullspace}
+
+        if appctx is None:
+            self.appctx = appctx_base
+        else:
+            self.appctx = {**appctx, **appctx_base}
+
+    @abstractmethod
+    def advance(self):
+        pass
+
+    @abstractmethod
+    def solver_stats(self):
+        pass
+
+
+# Stage derivative + stage value + (maybe?) RadauIIAIMEX
```

Comment:
I feel like we should remove the `(maybe?)` here before merging...

Reply:
This code clearly doesn't cover that `RadauIIAIMEX` case, though, so I guess we just remove that part of the comment entirely?

Reply:
I think it's not worth it for now, especially given our rather tentative embrace of the method.